package ru.practicum.shareit.exception;

import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpStatus;
import org.springframework.http.converter.HttpMessageNotReadableException;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestControllerAdvice;
import org.springframework.web.client.HttpServerErrorException;


@Slf4j
@RestControllerAdvice
public class ErrorHandler {

    @ExceptionHandler({ValidationException.class, MethodArgumentNotValidException.class,})
    @ResponseStatus(HttpStatus.BAD_REQUEST)
    public ErrorResponse notValidArgsException(ValidationException e) {
        log.info("Ошибка валидации: {}", e.getMessage());
        return new ErrorResponse(e.getMessage(), e.getMessage());
    }


    @ExceptionHandler({NotFoundException.class, HttpMessageNotReadableException.class,})
    @ResponseStatus(HttpStatus.NOT_FOUND)
    public ErrorResponse handleObjectNotFoundExc(NotFoundException e) {
        log.info("Элемент не найден {}", e.getMessage());
        return new ErrorResponse(e.getMessage(), e.getMessage());
    }

    @ExceptionHandler({HttpServerErrorException.InternalServerError.class})
    @ResponseStatus(HttpStatus.INTERNAL_SERVER_ERROR)
    public ErrorResponse handleOtherExc(Throwable e) {
        log.info("Ошибка 500 INTERNAL_SERVER_ERROR: ", e);
        return new ErrorResponse(e.getMessage(), e.getMessage());
    }

    @ExceptionHandler (DataConflict.class)
    @ResponseStatus(HttpStatus.CONFLICT)
    public ErrorResponse handleConflictExc(Throwable e) {
        log.info("Ошибка 409 CONFLICT: ", e);
        return new ErrorResponse(e.getMessage(), e.getMessage());
    }
}
